doc/tools/coqrst/coqdoc/main.py

```diff
@@ -48,28 +48,16 @@ def coqdoc(coq_code, coqdoc_bin=None):
     finally:
         os.remove(filename)
 
-def is_whitespace_string(elem):
-    return isinstance(elem, NavigableString) and elem.strip() == ""
-
-def strip_soup(soup, pred):
-    """Strip elements matching pred from front and tail of soup."""
-    while soup.contents and pred(soup.contents[-1]):
-        soup.contents.pop()
-
-    skip = 0
-    for elem in soup.contents:
-        if not pred(elem):
-            break
-        skip += 1
-
-    soup.contents[:] = soup.contents[skip:]
-
 def lex(source):
     """Convert source into a stream of (css_classes, token_string)."""
     coqdoc_output = coqdoc(source)
     soup = BeautifulSoup(coqdoc_output, "html.parser")
     root = soup.find(class_='code')
-    strip_soup(root, is_whitespace_string)
+    if root.children:
+        # strip the leading '\n'
+        first = next(root.children)
+        if isinstance(first, NavigableString) and first.string[0] == '\n':
+            first.string.replace_with(first.string[1:])
```

Comment:
I'm not sure we should drop that logic: we can't make assumptions about the way the elements returned by coqdoc will be nested.

Reply:
I do not really understand your comment. The logic I added is arguably a bit brittle, but I don't think it is any worse than what was there before:

<img width="195" alt="Screenshot 2020-05-08 at 13 19 52" src="https://user-images.githubusercontent.com/1117246/81401248-36f90000-912f-11ea-8332-c91f081592f5.png">

The new code significantly improves the appearance of `.. coqtop::` directives (see the generated documentation). Should coqdoc change its output, the only thing that might change compared to the previous code is that spacing looks a bit different.

Reply:
Sorry for the confusion: I like the change of style (of course! in fact, it's a bugfix, really ^^); I'm just wondering if there's a slightly better way to write it.

I guess what I'm suggesting boils down to finding the first NavigableString at any depth, rather than checking if the first element is a navigable string.  So maybe call next not on root.children but on a filtered version of root.contents containing NavigableStrings?